lib/anonymous-auth.js

```diff
@@ -0,0 +1,630 @@
+}
+
+/**
+ * Recursively freeze an object and all its properties
+ */
+function deepFreeze(obj) {
+    // Retrieve the property names defined on obj
+    Object.getOwnPropertyNames(obj).forEach(function (prop) {
+        var value = obj[prop];
+
+        // Freeze properties that are objects/arrays before freezing self
+        if (value && typeof (value) === 'object') {
+            deepFreeze(value);
+        }
+    });
+
+    return (Object.freeze(obj));
+}
+
+/**
+ * Anonymous access middleware
+ * Should be called before authentication for requests that might be public
+ */
+function anonymousAccessHandler(req, res, next) {
+    req.log.debug({
+        method: req.method,
+        path: req.path(),
+        hasAuth: !!(req.headers.authorization || req.headers.Authorization)
+    }, 'Checking for anonymous access');
+
+    // Check if anonymous access is enabled
+    if (!ANONYMOUS_ACCESS_CONFIG.enabled) {
+        req.log.debug('Anonymous access disabled by configuration');
+        next();
+        return;
+    }
+
+    // Production audit logging for anonymous access attempts
+    if (ANONYMOUS_ACCESS_CONFIG.auditAll) {
+        req.log.warn({
+            event: 'anonymous_access_attempt',
+            method: req.method,
+            path: req.path(),
+            userAgent: req.headers['user-agent'],
+            sourceIP: req.connection.remoteAddress,
+            timestamp: new Date().toISOString()
+        }, 'AUDIT: Anonymous access attempt');
+    }
+
+    // If request already has authentication, skip anonymous handling
+    if (req.headers.authorization || req.headers.Authorization) {
+        req.log.debug(
+            'Request has authentication headers, skipping anonymous access');
+        next();
+        return;
+    }
+
+    // Only allow GET/HEAD/OPTIONS operations for anonymous access
+    if (req.method !== 'GET' && req.method !== 'HEAD' &&
+        req.method !== 'OPTIONS') {
+        req.log.debug('Anonymous access only allowed for' +
+                      ' GET/HEAD/OPTIONS operations');
+        next();
+        return;
+    }
+
+    // Extract bucket name from request
+    var bucketName = extractBucketName(req);
+    if (!bucketName) {
+        req.log.debug('No bucket name found, skipping anonymous access');
+        next();
+        return;
+    }
+
+    // Mark this request as potentially anonymous and defer the decision
+    // The actual check will happen later when bucket metadata is available
+    req.log.debug({
+        bucket: bucketName,
+        method: req.method,
+        path: req.path()
+    }, 'Marking request for potential anonymous access - will verify later');
+
+    // Extract the account name from the path (/account/buckets/bucket)
+    var pathParts = req.path().split('/').filter(function (part) {
+        return (part.length > 0);
+    });
+    var accountName = pathParts[0]; // First part is the account name
+
+    // Set up anonymous context tentatively - this will be validated later
```

Comment:
Do we actually validate this later? `accountName` get's passed around to a few places but mostly _seem_ to be only for logging purposes so we may not need to do anything but I didn't see any validation taking place (or I have missed it.)

Reply:
No, accountName is not being validated down the stack; we only rely on the role metadata from the object of interest in the request. This is a security issue; I'll add a lookup to at least verify if the account exists.